File: app/src/main/java/com/example/cinemates/view/ui/details/movie/MovieInfoFragment.java
package com.example.cinemates.view.ui.details.movie;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.example.cinemates.adapter.ItemsRecyclerViewAdapter;
import com.example.cinemates.adapter.YoutubeVideoRecyclerViewAdapter;
import com.example.cinemates.databinding.FragmentMediaInfoBinding;
import com.example.cinemates.model.data.Movie;
import com.example.cinemates.model.data.Video;
import com.example.cinemates.util.ViewSize;
import com.example.cinemates.view.viewmodel.MovieViewModel;

import java.util.ArrayList;
import java.util.List;


public class MovieInfoFragment extends Fragment {

    private FragmentMediaInfoBinding mBinding;
    private MovieViewModel mViewModel;
    private ItemsRecyclerViewAdapter<Movie> mAdapter;
    private YoutubeVideoRecyclerViewAdapter mVideoAdapter;
    private Movie mMovie;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAdapter = new ItemsRecyclerViewAdapter<>(ViewSize.SMALL);
        mVideoAdapter = new YoutubeVideoRecyclerViewAdapter();
        mViewModel = new ViewModelProvider(getActivity()).get(MovieViewModel.class);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mBinding = FragmentMediaInfoBinding.inflate(inflater, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mMovie = (Movie) getArguments().getSerializable("movie");
        mBinding.recommendedRecyclerView.setAdapter(mAdapter);
        mBinding.recommendedRecyclerView.setEmptyView(mBinding.emptyViewRecommended.getRoot());
        mBinding.videosRecyclerView.setAdapter(mVideoAdapter);

        observe();
        mViewModel.getMovieDetails(mMovie.getId());
        mViewModel.getSimilar(mMovie.getId());
        mViewModel.getVideos(mMovie.getId());

        mBinding.collectionName.collectionName.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                FragmentManager fm = getActivity().getSupportFragmentManager();
                CollectionDialogFragment collectionDialogFragment = CollectionDialogFragment.newInstance(mMovie.getBelongs_to_collection());
                collectionDialogFragment.show(fm, "fragment_show_collection");
            }
        });

        mBinding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                // TODO: 22/08/2022 add navigation to DetailActorFragment (next feature required new branch)
               /* MovieDetailsFragmentDirections.ActionDetailMediaContentFragmentToDetailActorMediaFragment action =
                        MovieDetailsFragmentDirections.actionDetailMediaContentFragmentToDetailActorMediaFragment(mMovie);
                Navigation.findNavController(view).navigate(action);*/

            }
        });


    }

    private void observe() {
        mViewModel.getMovie().observe(getViewLifecycleOwner(), new Observer<Movie>() {
            @Override
            public void onChanged(Movie movie) {
                mBinding.setMovie(movie);
                mMovie = movie;
            }
        });
        mViewModel.getMovieSimilar().observe(getViewLifecycleOwner(), new Observer<List<Movie>>() {
            @Override
            public void onChanged(List<Movie> movies) {
                mAdapter.addItems(movies);
            }
        });
        mViewModel.getMovieVideos().observe(getViewLifecycleOwner(), new Observer<List<Video>>() {
            @Override
            public void onChanged(List<Video> videos) {
                if (!videos.isEmpty()) {
                    mBinding.textSectionVideo.setVisibility(View.VISIBLE);
                    mVideoAdapter.setDataList((ArrayList<Video>) videos);
                } else {
                    mBinding.textSectionVideo.setVisibility(View.GONE);
                }
            }
        });
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mBinding = null;

    }


}

File: app/src/main/java/com/example/cinemates/view/ui/details/movie/MovieInfoFragment.kt
package com.example.cinemates.view.ui.details.movie

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import com.example.cinemates.R
import com.example.cinemates.adapter.MultiViewTypeRecyclerViewAdapter
import com.example.cinemates.adapter.YoutubeVideoRecyclerViewAdapter
import com.example.cinemates.databinding.FragmentMovieInfoBinding
import com.example.cinemates.model.data.Movie
import com.example.cinemates.util.ViewSize

class MovieInfoFragment : Fragment() {

    private var _binding: FragmentMovieInfoBinding? = null
    private val binding: FragmentMovieInfoBinding
        get() = _binding!!
    private lateinit var similarAdapter: MultiViewTypeRecyclerViewAdapter<Movie>
    private lateinit var movieIntoCollectionAdapter: MultiViewTypeRecyclerViewAdapter<Movie>
    private lateinit var videoAdapter: YoutubeVideoRecyclerViewAdapter
    private val viewModel: MovieDetailsViewModel by activityViewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        similarAdapter = MultiViewTypeRecyclerViewAdapter(ViewSize.SMALL)
        videoAdapter = YoutubeVideoRecyclerViewAdapter()
        movieIntoCollectionAdapter = MultiViewTypeRecyclerViewAdapter(ViewSize.SMALL)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding = FragmentMovieInfoBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.apply {

            recommendedRecyclerView.adapter = similarAdapter
            recommendedRecyclerView.setEmptyView(emptyViewRecommended.root)
            videosRecyclerView.adapter = videoAdapter

            collectionView.apply {
                moviesIntoCollection.adapter = movieIntoCollectionAdapter
                collectionCover.coverTitle.setOnClickListener {
                    //Remove collection cover
                    collectionCover.root.visibility = View.INVISIBLE

                }
                contentTitle.setOnClickListener {
                    //Add collection cover
                    collectionCover.root.visibility = View.VISIBLE
                }
            }


            fab.setOnClickListener {
                //Open bottomSheetFragment
                findNavController().navigate(R.id.action_movieDetailsFragment_to_bottomSheetFragment)
            }

            viewModel.selectedMovie.observe(viewLifecycleOwner) { selectedMovie ->
                movie = selectedMovie
            }
            viewModel.videos.observe(viewLifecycleOwner) { videos ->
                if (videos.isEmpty()) {
                    textSectionVideo.visibility = View.GONE
                    videosRecyclerView.visibility = View.GONE
                } else {
                    textSectionVideo.visibility = View.VISIBLE
                    videosRecyclerView.visibility = View.VISIBLE
                    videoAdapter.setDataList(videos)
                }
            }

        }

        viewModel.similarMovies.observe(viewLifecycleOwner) { similarMovies ->
            similarAdapter.addItems(similarMovies.toMutableList())
        }
        viewModel.moviesBelongsCollection.observe(viewLifecycleOwner) { moviesBelongsCollection ->
            movieIntoCollectionAdapter.addItems(moviesBelongsCollection.toMutableList())
        }

    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }


}